File: CS/DXApplication7/CustomPictureEdit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System.Collections;

namespace DXApplication7 {
    [UserRepositoryItem("RegisterCustomPictureEdit")]
    public class RepositoryItemCustomPictureEdit : RepositoryItemPictureEdit {
        static RepositoryItemCustomPictureEdit() {
            RegisterCustomPictureEdit();
        }
        public const string CustomEditName = "CustomPictureEdit";
        public RepositoryItemCustomPictureEdit() {
            HighlightFigureColor = Color.FromArgb(100, Color.Blue);
            ActiveFigureColor = Color.FromArgb(100, Color.Orange);
        }
        public override string EditorTypeName {
            get {
                return CustomEditName;
            }
        }
        public static void RegisterCustomPictureEdit() {
            Image img = null;
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName, typeof(CustomPictureEdit), typeof(RepositoryItemCustomPictureEdit), typeof(CustomPictureEditViewInfo), new CustomPictureEditPainter(), true, img));
        }
        Color highlightFigureColor;
        public Color HighlightFigureColor {
            get { return highlightFigureColor; }
            set {
                if (highlightFigureColor != value) {
                    highlightFigureColor = value;
                    OnPropertiesChanged();
                }
            }
        }
        Color activeFigureColor;
        public Color ActiveFigureColor {
            get { return activeFigureColor; }
            set {
                if (activeFigureColor != value) {
                    activeFigureColor = value;
                    OnPropertiesChanged();
                }
            }
        }
        public override void Assign(RepositoryItem item) {
            BeginUpdate();
            try {
                base.Assign(item);
                RepositoryItemCustomPictureEdit source = item as RepositoryItemCustomPictureEdit;
                if (source == null) return;
                HighlightFigureColor = source.HighlightFigureColor;
                ActiveFigureColor = source.ActiveFigureColor;
            }
            finally {
                EndUpdate();
            }
        }
    }

    [ToolboxItem(true)]
    public class CustomPictureEdit : PictureEdit {
        static CustomPictureEdit() {
            RepositoryItemCustomPictureEdit.RegisterCustomPictureEdit();
        }
        public CustomPictureEdit() { }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemCustomPictureEdit Properties {
            get {
                return base.Properties as RepositoryItemCustomPictureEdit;
            }
        }
        public override string EditorTypeName {
            get {
                return RepositoryItemCustomPictureEdit.CustomEditName;
            }
        }
        NestedFigureCollection nestedFigures;
        public NestedFigureCollection NestedFigures {
            get {
                if (nestedFigures == null) {
                    nestedFigures = new NestedFigureCollection();
                }
                return nestedFigures;
            }
            set {
                if (nestedFigures != value) {
                    nestedFigures = value;
                    OnPropertiesChanged();
                }
            }
        }
        protected override void OnMouseMove(System.Windows.Forms.MouseEventArgs e) {
            base.OnMouseMove(e);
            UpdateNestedFigures(e.Location);
        }
        protected override void OnMouseDown(System.Windows.Forms.MouseEventArgs e) {
            base.OnMouseDown(e);
            ActivateNestedFigures(e.Location);
        }
        protected void UpdateNestedFigures(Point pt) {
            if (NestedFigures.Count > 0) {
                Point imagePoint = ViewportToImage(pt);
                foreach (Figure figure in NestedFigures) {
                    figure.Selected = figure.SourceRect.Contains(imagePoint);
                }
                Invalidate();
            }
        }
        protected void ActivateNestedFigures(Point pt) {
            if (NestedFigures.Count > 0) {
                Point imagePoint = ViewportToImage(pt);
                foreach (Figure figure in NestedFigures) {
                    if(figure.SourceRect.Contains(imagePoint)){
                        figure.Active = !figure.Active;
                        break;
                    }
                }
                Invalidate();
            }
        }
        internal protected Rectangle FigureRectToEditorRect(Figure figure){
            return new Rectangle(ImageToViewport(figure.SourceRect.Location), figure.SourceRect.Size);
        }
         public Image GetImageWithFigures() {
             if (Image == null) return null;
             if(NestedFigures.Count == 0) return Image;
             Image newImage = Image.Clone() as Image;
             using(Graphics graphics = Graphics.FromImage(newImage))
                foreach (Figure figure in NestedFigures) {
                    if(figure.Active){
                        using (SolidBrush brush = new SolidBrush(Properties.ActiveFigureColor)) {
                            graphics.FillRectangle(brush, figure.SourceRect);
                        }
                    }
                }
             return newImage;
        }
    }

    public class CustomPictureEditViewInfo : PictureEditViewInfo {
        public CustomPictureEditViewInfo(RepositoryItem item)
            : base(item) {
        }
    }
    public class CustomPictureEditPainter : PictureEditPainter {
        public CustomPictureEditPainter() {
        }
        protected override void DrawImage(ControlGraphicsInfoArgs info) {
            base.DrawImage(info);
            PictureEditViewInfo viewInfo = info.ViewInfo as PictureEditViewInfo;
            if (viewInfo.OwnerEdit != null) {
                CustomPictureEdit edit = viewInfo.OwnerEdit as CustomPictureEdit;
                if (edit.NestedFigures.Count > 0) {
                    foreach (Figure figure in edit.NestedFigures) {
                        if (!figure.Active && !figure.Selected) continue;
                        Rectangle rect = edit.FigureRectToEditorRect(figure);
                        info.Cache.FillRectangle(figure.Selected ? edit.Properties.HighlightFigureColor : edit.Properties.ActiveFigureColor, rect);
                    }
                }
            }
        }
    }
    public class Figure {
        public Rectangle SourceRect { get; set; }
        internal bool Active { get; set; }
        internal bool Selected { get; set; }
    }
    public class NestedFigureCollection : CollectionBase{
        public void Add(Rectangle figureRect) {
            List.Add(new Figure() { SourceRect = figureRect});
        }
        public void AddRange(IEnumerable figureRects) {
            foreach (Rectangle rect in figureRects) {
                List.Add(new Figure() { SourceRect = rect });
            }
        }
        public void Remove(Rectangle figureRect) {
            var figure = List.OfType<Figure>().FirstOrDefault(f => f.SourceRect == figureRect);
            if (figure != null) {
                List.Remove(figure);
            }
        }
        public Figure this[int index] {
            get { return List[index] as Figure; }
        }
    }
}

File: VB/DXApplication7/CustomPictureEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System.Collections

Namespace DXApplication7
    <UserRepositoryItem("RegisterCustomPictureEdit")> _
    Public Class RepositoryItemCustomPictureEdit
        Inherits RepositoryItemPictureEdit

        Shared Sub New()
            RegisterCustomPictureEdit()
        End Sub
        Public Const CustomEditName As String = "CustomPictureEdit"
        Public Sub New()
            HighlightFigureColor = Color.FromArgb(100, Color.Blue)
            ActiveFigureColor = Color.FromArgb(100, Color.Orange)
        End Sub
        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return CustomEditName
            End Get
        End Property
        Public Shared Sub RegisterCustomPictureEdit()
            Dim img As Image = Nothing
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(CustomPictureEdit), GetType(RepositoryItemCustomPictureEdit), GetType(CustomPictureEditViewInfo), New CustomPictureEditPainter(), True, img))
        End Sub

        Private highlightFigureColor_Renamed As Color
        Public Property HighlightFigureColor() As Color
            Get
                Return highlightFigureColor_Renamed
            End Get
            Set(ByVal value As Color)
                If highlightFigureColor_Renamed <> value Then
                    highlightFigureColor_Renamed = value
                    OnPropertiesChanged()
                End If
            End Set
        End Property

        Private activeFigureColor_Renamed As Color
        Public Property ActiveFigureColor() As Color
            Get
                Return activeFigureColor_Renamed
            End Get
            Set(ByVal value As Color)
                If activeFigureColor_Renamed <> value Then
                    activeFigureColor_Renamed = value
                    OnPropertiesChanged()
                End If
            End Set
        End Property
        Public Overrides Sub Assign(ByVal item As RepositoryItem)
            BeginUpdate()
            Try
                MyBase.Assign(item)
                Dim source As RepositoryItemCustomPictureEdit = TryCast(item, RepositoryItemCustomPictureEdit)
                If source Is Nothing Then
                    Return
                End If
                HighlightFigureColor = source.HighlightFigureColor
                ActiveFigureColor = source.ActiveFigureColor
            Finally
                EndUpdate()
            End Try
        End Sub
    End Class

    <ToolboxItem(True)> _
    Public Class CustomPictureEdit
        Inherits PictureEdit

        Shared Sub New()
            RepositoryItemCustomPictureEdit.RegisterCustomPictureEdit()
        End Sub
        Public Sub New()
        End Sub
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
        Public Shadows ReadOnly Property Properties() As RepositoryItemCustomPictureEdit
            Get
                Return TryCast(MyBase.Properties, RepositoryItemCustomPictureEdit)
            End Get
        End Property
        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return RepositoryItemCustomPictureEdit.CustomEditName
            End Get
        End Property

        Private nestedFigures_Renamed As NestedFigureCollection
        Public Property NestedFigures() As NestedFigureCollection
            Get
                If nestedFigures_Renamed Is Nothing Then
                    nestedFigures_Renamed = New NestedFigureCollection()
                End If
                Return nestedFigures_Renamed
            End Get
            Set(ByVal value As NestedFigureCollection)
                If nestedFigures_Renamed IsNot value Then
                    nestedFigures_Renamed = value
                    OnPropertiesChanged()
                End If
            End Set
        End Property
        Protected Overrides Sub OnMouseMove(ByVal e As System.Windows.Forms.MouseEventArgs)
            MyBase.OnMouseMove(e)
            UpdateNestedFigures(e.Location)
        End Sub
        Protected Overrides Sub OnMouseDown(ByVal e As System.Windows.Forms.MouseEventArgs)
            MyBase.OnMouseDown(e)
            ActivateNestedFigures(e.Location)
        End Sub
        Protected Sub UpdateNestedFigures(ByVal pt As Point)
            If NestedFigures.Count > 0 Then
                Dim imagePoint As Point = ViewportToImage(pt)
                For Each figure As Figure In NestedFigures
                    figure.Selected = figure.SourceRect.Contains(imagePoint)
                Next figure
                Invalidate()
            End If
        End Sub
        Protected Sub ActivateNestedFigures(ByVal pt As Point)
            If NestedFigures.Count > 0 Then
                Dim imagePoint As Point = ViewportToImage(pt)
                For Each figure As Figure In NestedFigures
                    If figure.SourceRect.Contains(imagePoint) Then
                        figure.Active = Not figure.Active
                        Exit For
                    End If
                Next figure
                Invalidate()
            End If
        End Sub
        Protected Friend Function FigureRectToEditorRect(ByVal figure As Figure) As Rectangle
            Return New Rectangle(ImageToViewport(figure.SourceRect.Location), figure.SourceRect.Size)
        End Function
         Public Function GetImageWithFigures() As Image
             If Image Is Nothing Then
                 Return Nothing
             End If
             If NestedFigures.Count = 0 Then
                 Return Image
             End If
             Dim newImage As Image = TryCast(Image.Clone(), Image)
             Using graphics As Graphics = System.Drawing.Graphics.FromImage(newImage)
                For Each figure As Figure In NestedFigures
                    If figure.Active Then
                        Using brush As New SolidBrush(Properties.ActiveFigureColor)
                            graphics.FillRectangle(brush, figure.SourceRect)
                        End Using
                    End If
                Next figure
             End Using
             Return newImage
         End Function
    End Class

    Public Class CustomPictureEditViewInfo
        Inherits PictureEditViewInfo

        Public Sub New(ByVal item As RepositoryItem)
            MyBase.New(item)
        End Sub
    End Class
    Public Class CustomPictureEditPainter
        Inherits PictureEditPainter

        Public Sub New()
        End Sub
        Protected Overrides Sub DrawImage(ByVal info As ControlGraphicsInfoArgs)
            MyBase.DrawImage(info)
            Dim viewInfo As PictureEditViewInfo = TryCast(info.ViewInfo, PictureEditViewInfo)
            If viewInfo.OwnerEdit IsNot Nothing Then
                Dim edit As CustomPictureEdit = TryCast(viewInfo.OwnerEdit, CustomPictureEdit)
                If edit.NestedFigures.Count > 0 Then
                    For Each figure As Figure In edit.NestedFigures
                        If Not figure.Active AndAlso Not figure.Selected Then
                            Continue For
                        End If
                        Dim rect As Rectangle = edit.FigureRectToEditorRect(figure)
                        info.Cache.FillRectangle(If(figure.Selected, edit.Properties.HighlightFigureColor, edit.Properties.ActiveFigureColor), rect)
                    Next figure
                End If
            End If
        End Sub
    End Class
    Public Class Figure
        Public Property SourceRect() As Rectangle
        Friend Property Active() As Boolean
        Friend Property Selected() As Boolean
    End Class
    Public Class NestedFigureCollection
        Inherits CollectionBase

        Public Sub Add(ByVal figureRect As Rectangle)
            List.Add(New Figure() With {.SourceRect = figureRect})
        End Sub
        Public Sub AddRange(ByVal figureRects As IEnumerable)
            For Each rect As Rectangle In figureRects
                List.Add(New Figure() With {.SourceRect = rect})
            Next rect
        End Sub
        Public Sub Remove(ByVal figureRect As Rectangle)
            Dim figure = List.OfType(Of Figure)().FirstOrDefault(Function(f) f.SourceRect = figureRect)
            If figure IsNot Nothing Then
                List.Remove(figure)
            End If
        End Sub
        Default Public ReadOnly Property Item(ByVal index As Integer) As Figure
            Get
                Return TryCast(List(index), Figure)
            End Get
        End Property
    End Class
End Namespace
